src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -1453,10 +1453,15 @@ void GcsActorManager::RestartActor(const ActorID &actor_id,
       << ", death context type = " << GetActorDeathCauseString(death_cause)
       << ", remaining_restarts = " << remaining_restarts;
 
-  if (remaining_restarts != 0) {
+  if (remaining_restarts != 0 ||
+      (need_reschedule && max_restarts > 0 && mutable_actor_table_data->preempted())) {
     // num_restarts must be set before updating GCS, or num_restarts will be inconsistent
     // between memory cache and storage.
-    mutable_actor_table_data->set_num_restarts(num_restarts + 1);
+    if (mutable_actor_table_data->preempted()) {
+      mutable_actor_table_data->set_preempted(false);
```

Comment:
This logic makes the method not idempotent. Do we expect that it's called exactly once per actor death?

Reply:
I just found that the newly added check for idempotency is causing GCS to crash with the following assertion error:
```
[2025-07-28 17:27:49,785 C 73159 8451788] (gcs_server) gcs_actor_scheduler.cc:237:  Check failed: node_it != node_to_actors_when_leasing_.end()
```
I am still figuring out why, but I have already reverted the check first.